http.go

```diff
@@ -106,6 +117,45 @@ func InstrumentHandlerFunc(label string, f func(http.ResponseWriter, *http.Reque
 	return InstrumentHandler(label, http.HandlerFunc(f))
 }
 
+// traceSampler is a time-based request sampler for tracing.
+//
+// It allows at most one request to be traced during each time interval.
+type traceSampler struct {
+	intervalNs float64
+
+	mu               sync.Mutex
+	rng              *rand.Rand
+	nextSampleTimeNs atomic.Int64
+}
+
+// newTraceSampler returns a new traceSampler that allows at most one request
+// to be traced during each time interval.
+func newTraceSampler(interval time.Duration) *traceSampler {
+	return &traceSampler{
+		intervalNs: float64(interval / time.Nanosecond),
+		rng:        rand.New(rand.NewSource(time.Now().UnixNano())),
+	}
+}
+
+// shouldTrace returns true iff a request should be traced.
+func (s *traceSampler) shouldTrace() bool {
+	now := time.Now().UnixNano()
```

Comment:
Add a comment:
  // Fast path check avoids acquiring a mutex or modifying the sampler in any way.